### PR TITLE
MEPTS-158 TX_PVLS Consider most recent state if pregnant and breastfeeding

### PR DESCRIPTION
**Pull request by ningosi**
This initial work on the MEPTS-158, more improvements ongoing and adding/fixing failing tests

### CODE REVIEW

File: api/src/test/java/org/openmrs/module/eptsreports/reporting/intergrated/calculation/pvls/BreastfeedingPregnantCalculationTest.java [hunk 1]
```diff
@@ -66,10 +68,11 @@ public void initialise() throws Exception {
   @Test
   public void shouldEvaluateOnePatientMarkedAsPregnantByEnrollInPTVProgram() {
```
Comment (ynurmahomed) [inline api/src/test/java/org/openmrs/module/eptsreports/reporting/intergrated/calculation/pvls/BreastfeedingPregnantCalculationTest.java]: These scenarios belong to `BreastfeedingDateCalculation`. Please move to a file named `BreastfeedingDateCalculationTest` and adjust assertions accordingly.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingPregnantCalculation.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        calculate(
+            Context.getRegisteredComponents(BreastfeedingDateCalculation.class).get(0),
+            femaleCohort,
+            context);
+    for (Integer ptId : femaleCohort) {
+      boolean isCandidate = false;
+      Date pregnancyDate = (Date) pregnantDateMap.get(ptId).getValue();
+      Date breastfeedingDate = (Date) breastfeedingDateMap.get(ptId).getValue();
+      if (state.equals(PregnantOrBreastfeedingWomen.PREGNANTwOMEN)) {
+        if (pregnancyDate != null) {
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingPregnantCalculation.java]: This block can be simplified as:
```
if (pregnancyDate != null) {
  isCandidate = breastfeedingDate == null || breastfeedingDate.before(pregnancyDate);
}
```
Similarly for when `state.equals(PregnantOrBreastfeedingWomen.BREASTFEEDINGWOMEN)`.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingPregnantCalculation.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    PregnantOrBreastfeedingWomen state =
+        (PregnantOrBreastfeedingWomen) parameterValues.get("state");
+
+    // get female patients only
+    Set<Integer> femaleCohort = EptsCalculationUtils.female(cohort, context);
+
+    CalculationResultMap pregnantDateMap =
+        calculate(
+            Context.getRegisteredComponents(PregnantDateCalculation.class).get(0),
```
[the remark this answers is not in the document]
Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingPregnantCalculation.java]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingDateCalculation.java [hunk 4]
```diff
@@ -29,7 +31,7 @@
  * @return CalculationResultMap
  */
 @Component
-public class BreastfeedingCalculation extends AbstractPatientCalculation {
+public class BreastfeedingDateCalculation extends AbstractPatientCalculation {
 
   @Autowired private HivMetadata hivMetadata;
```
[the remark this answers is not in the document]
Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingDateCalculation.java]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingPregnantCalculation.java (hunk 2 above)
Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/pvls/BreastfeedingPregnantCalculation.java]: Done

File: api/src/test/java/org/openmrs/module/eptsreports/reporting/intergrated/calculation/pvls/BreastfeedingPregnantCalculationTest.java (hunk 1 above)
Comment (ningosi) [reply api/src/test/java/org/openmrs/module/eptsreports/reporting/intergrated/calculation/pvls/BreastfeedingPregnantCalculationTest.java]: For pregnant patients, one of the option is to check the PTV program, how is that breastfeeding?

Comment (ningosi) [reply api/src/test/java/org/openmrs/module/eptsreports/reporting/intergrated/calculation/pvls/BreastfeedingPregnantCalculationTest.java]: In any case it should be **PregnantDateCalculationTest** and NOT **BreastfeedingDateCalculationTest**

Comment (ynurmahomed) [reply api/src/test/java/org/openmrs/module/eptsreports/reporting/intergrated/calculation/pvls/BreastfeedingPregnantCalculationTest.java]: That is what I meant to say

**Comment by ningosi** [comment]
@ynurmahomed  Reviews addressed. Find time to go through again

File: api/src/test/java/org/openmrs/module/eptsreports/reporting/intergrated/calculation/pvls/BreastfeedingPregnantCalculationTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  @Before
+  public void initialise() throws Exception {
+    executeDataSet("pvlsTest.xml");
+    params.put("state", PregnantOrBreastfeedingWomen.PREGNANTWOMEN);
+  }
+
+  @Test
+  public void shouldReturnPregnantIfMostRecent() {
+    setEvaluationContext(testsHelper.getDate("2018-09-20 00:00:00.0"));
+    params.put("state", PregnantOrBreastfeedingWomen.BREASTFEEDINGWOMEN);
```
Comment (ynurmahomed) [inline api/src/test/java/org/openmrs/module/eptsreports/reporting/intergrated/calculation/pvls/BreastfeedingPregnantCalculationTest.java]: @ningosi shouldn't it be `PREGNANTWOMEN` here? Maybe this test belongs to `shouldReturnBreastfeedingIfMostRecent`.

Comment (ningosi) [reply api/src/test/java/org/openmrs/module/eptsreports/reporting/intergrated/calculation/pvls/BreastfeedingPregnantCalculationTest.java]: Sure, I will turn this around

**Pull request merged**